fe/fe-core/src/main/java/com/starrocks/connector/RemoteFileOperations.java

```diff
@@ -136,6 +139,80 @@ public List<RemoteFileInfo> getRemoteFiles(List<Partition> partitions, Options o
+        RemoteFileScanContext scanContext = new RemoteFileScanContext(table);
+
+        // compute cache miss size.
+        {
+            List<RemotePathKey> remotePathKeys = new ArrayList<>();
+            for (Partition partition : partitions) {
+                RemotePathKey key = RemotePathKey.of(partition.getFullPath(), isRecursive);
+                remotePathKeys.add(key);
+            }
+
+            int cacheMissSize = partitions.size();
+            if (enableCatalogLevelCache && params.isUseCache()) {
+                cacheMissSize = cacheMissSize - remoteFileIO.getPresentRemoteFiles(remotePathKeys).size();
+            }
+            Tracers.count(Tracers.Module.EXTERNAL, HMS_PARTITIONS_REMOTE_FILES, cacheMissSize);
+        }
+
+        class MyAsyncTaskQueue extends AsyncTaskQueue<RemoteFileInfo> implements RemoteFileInfoSource {
+            public MyAsyncTaskQueue(Executor executor) {
+                super(executor);
+            }
+
+            @Override
+            public List<RemoteFileInfo> getOutputs(int maxSize) {
+                try (Timer ignored = Tracers.watchScope(Tracers.Module.EXTERNAL, HMS_PARTITIONS_REMOTE_FILES)) {
+                    return super.getOutputs(maxSize);
+                }
+            }
+
+            @Override
+            public int computeOutputSize(RemoteFileInfo output) {
+                List<RemoteFileDesc> files = (output.getFiles());
+                int size = 1;
+                if (files != null) {
+                    size = Math.max(size, files.size());
+                }
+                return size;
+            }
+
+            @Override
+            public RemoteFileInfo getOutput() {
+                List<RemoteFileInfo> res = getOutputs(1);
+                Preconditions.checkArgument(res.size() == 1);
+                return res.get(0);
+            }
+        }
+
+        SessionVariable sv = SessionVariable.DEFAULT_SESSION_VARIABLE;
+        if (ConnectContext.get() != null) {
+            sv = ConnectContext.get().getSessionVariable();
+        }
+        MyAsyncTaskQueue asyncTaskQueue = new MyAsyncTaskQueue(pullRemoteFileExecutor);
+        asyncTaskQueue.setMaxOutputQueueSize(sv.getConnectorRemoteFileAsyncQueueSize());
+        asyncTaskQueue.setMaxRunningTaskCount(sv.getConnectorRemoteFileAsyncTaskSize());
+        List<AsyncTaskQueue.Task<RemoteFileInfo>> tasks = new ArrayList<>();
+        List attachments = params.getPartitionAttachments();
+        for (int i = 0; i < partitions.size(); i++) {
+            final Partition partition = partitions.get(i);
+            final RemotePathKey pathKey = RemotePathKey.of(partition.getFullPath(), isRecursive);
+            final Object attachment = (attachments != null) ? attachments.get(i) : null;
+            pathKey.setScanContext(scanContext);
+            tasks.add(() -> {
+                Map<RemotePathKey, List<RemoteFileDesc>> res = remoteFileIO.getRemoteFiles(pathKey);
```

Comment:
async more:
1. get partition
2. prune partition
3. list files of this partition